File: js/653.js

var depthFirstIterate = function (root, k, nums) {
  
  if (!root || nums[0] + nums[nums.length - 1] >= k) {
    return
  }

  depthFirstIterate(root.left, k, nums)
  nums.push(root.val)
  depthFirstIterate(root.right, k, nums)
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) {
  
  var nums = []
  depthFirstIterate(root, k, nums)
   
  for(var i = 0; i < nums.length; i++) {
    for(var j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === k) {
        return true
      }
    }
  }

  return false
};

var tree = {"val":5,"right":{"val":6,"right":{"val":7,"right":null,"left":null},"left":null},"left":{"val":3,"right":{"val":4,"right":null,"left":null},"left":{"val":2,"right":null,"left":null}}}
var k = 9
console.log(findTarget(tree, k), true);
var k = 100
console.log(findTarget(tree, k), false);

File: py/653.py
def dfs(root, result, d):

    if not root:
        result.append(str(None))
        return

    start = len(result)
    result.append(str(root.val))
    dfs(root.left, result, d)
    dfs(root.right, result, d)
    end = len(result) + 1
    k = ','.join(result[start:end])
    if k in d:
        d[k] += 1
    else:
        d[k] = 1

def buildTree(nodes):
    val = nodes.pop()
    if val == None:
        return None
    root = TreeNode(val)
    root.left = buildTree(nodes)
    root.right = buildTree(nodes)
    return root

def parse(s):
    nodes = list(map(lambda x: int(x) if x != 'None' else None, s.split(',')))
    nodes.reverse()
    return buildTree(nodes)


class Solution:
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """
        result=[]
        d={}
        # d2={}
        dfs(root, result, d)
        result2=[]
        for (k, v) in d.items():
            # print("k:%s, v:%s" % (k, v))
            if v > 1:
                # result2.append(d2[k])
                print("k:%s, v:%s" % (k, v))
                result2.append(parse(k))
        return result2

class TreeNode:
    def __init__(self, val):
        self.val=val
        self.left=self.right=None
